File: ClassReader.py
# This Python file uses the following encoding: utf-8

import os
from ClassStorer import ClassStorer
from collections import OrderedDict
from specialvals import *


class ClassReader:
    def __init__(self):
        pass


    def get_classes(self):
        return OrderedDict(sorted(self.classes.items())) # return sorted dict but keep unsorted in original


    def readClasses(self):
        self.classes = dict()

        classCodeFiles = [file for file in os.listdir(ClassStorer.export_dir) if ".cpp" in file]
        for file in classCodeFiles:
            class_name = file.split(".")[0] # remove file extension to get class name for now
            self.classes[class_name] = dict()

            with open(ClassStorer.export_dir + "/" + file) as fr:
                func_body = False
                cur_func_name = None
                for line in fr:
                    if class_name + "::" in line:
                        cur_func_name = line.split("::")[1].split("(")[0]
                        self.classes[class_name][cur_func_name] = []
                        func_body = True
                    elif func_body:
                        self.classes[class_name][cur_func_name].append(line)
                        if line.rstrip("\n") == "}":
                            cur_func_name = None
                            func_body = False

File: CPP_Proj/classreader.cpp
#include "classreader.h"
#include "classstorer.h"

void ClassReader::readClasses()
{
  this->classes.clear();

  for (const auto & entry : fs::directory_iterator(ClassStorer::ExportDir.toStdString().c_str()))
  {
    QString fileP = QString(entry.path().generic_string().c_str());
    if (fileP.endsWith(".cpp"))
    {
        QString className = fileP.split('.')[0]; // remove file extension to get class name for now
        this->classes.insert_or_assign(className, FixedClass(className));

        QFile file(fileP);
        if(!file.open(QIODevice::ReadOnly)) {
            cout << "ERROR: File could not be opened - " << fileP.toStdString().c_str() << endl;
        }

        bool funcBody = false;
        QString curFuncName = nullptr;
        FixedFunction curFunc(nullptr);
        QTextStream in(&file);
        while(!in.atEnd()) {
            QString line = in.readLine().remove(QRegExp("([\n]{1})$"));
            if (line.contains(className + "::")) {
                curFuncName = line.split("::")[1].split("(")[0];
                curFunc = FixedFunction(curFuncName);
                funcBody = true;
            } else if (funcBody) {
                curFunc.addCodeLine(line);
                if (line.remove(QRegExp("([\n]+)$")) == QString("}")) {
                    this->classes[className].addFunction(curFunc);
                    curFuncName = nullptr;
                    funcBody = false;
                }
            }
        }
        file.close();
    }
  }
}
